### DISCUSSION TITLE
Add registration interface to template helper

### DISCUSSION THREAD

**Discussion by RoboMagus** [home-assistant/architecture#933] [General]
# Context
Functions available in template environments are limited to those available in the template helper.
For components to provide a template function, the template helper needs to "know" about this component. This would introduce a dependency issue that needs to be avoided.

Usecases for these template function include data that is usefull for automations and scripts but not best represented by entities. E.g. a list of current issues in the `issue_registry`.

## Registation interface
To enable expansion of the functions available in the template context some form of registration interface is required that allows components that would expose data to template environments to register this function without creating this unwanted dependency.

# Proposal
Implement such a registration interface. 
Using the solution proposed in home-assistant/core#94113 as a reference, or a wildly different implementation with similar effect.

## Goals
Provide a straightforward interface for other components to use that enable the expansion of functions available in template environments.

## Problems and Pitfalls
Performance. (Re-)Creating all global template functions / filters with the addition of (many) more globals could become too expensive.

# Alternatives
Limit functions available in the template environment to the `helpers` scope, as has been the case. Additional functionality must then be exposed as a helper and can then be added to the template environment.

**Comment by frenck** [comment]
Please don't use a PR to reference your intentions, instead, write up a proposal (see, for example #929). We don't want to end up reviewing a PR here, we want to discuss a proposal, with the possible use cases and the possible solutions.

Describe the context, set out expectations, use cases, the proposed solution, and its technical implications; maybe alternative you've considered and possible problems/risks you might be able to foresee.

Thanks 👍 

../Frenck

**Comment by allenporter** [comment]
Previously posted a proposal in another discussion thread https://github.com/home-assistant/architecture/discussions/777#discussioncomment-3080268 about adding templates to integrations, but it was focused on I/O. I've rewritten the proposal to exclude the I/O case below.

I think it would make sense if you proposed some concrete use cases to show this could be useful for a few different components/integrations.  The original motivation I had for async and I/O was solved with aligning on the services with response data proposal, so I personally do not have a use case for this at the moment.

---

I've only limited experience with templates, but I think the main considerations are:
- Integration must be isolated from each other and other template globals
- Only specific functions are allow-listed.
- Return values should also be limited in scope (e.g. simple values, dataclass with a finite set of attributes, etc)

So the simplest possible API I can think of is for some `DOMAIN` have a platform `template.py`:
```
# Allow listed set of functions exported as `DOMAIN.<function name>(...)`
TEMPLATE_FUNCTIONS = {
    "base64_encode",
}

def base64_encode(hass: HomeAssistant, value: str) -> str:
    """Perform base64 encode."""
    return base64.b64encode(value.encode("utf-8")).decode("utf-8")
```

So then from a template we can call these like:
-`{{ my_domain.base64_encode("example") }}`

We would only want to support function calls, and not exposing other object types.  New template functions would not be a replacement for other existing ways to use entities. There are many concerns with using async I/O from templates so that would not be allowed.

**Comment by balloob** [comment]
I suggest to put this architecture issue on hold and re-evaluated no earlier than in 6 months after we see the full impact of service return values that are coming in HA 2023.7.

**Comment by SLaks** [comment]
This came up in https://github.com/home-assistant/core/pull/102628, for a couple of reasons:

 - I want to expose a Python [`HDate` instance](https://github.com/py-libhdate/py-libhdate/blob/main/hdate/date.py), not a JSON-serializable object, for use in templates.
   - `HDate` has a number of properties that are expensive to compute, so I don't want to eagerly serialize it to JSON from a service call.
   - It also has [recursive properties](https://github.com/py-libhdate/py-libhdate/blob/31f4e8e4af622212f9964cd1e0f558e5f7e5803f/hdate/date.py#L259-L267).
 - I want to be able to get `HDate`s from inside templates, so that I can loop over a series of dates, get a date for a computed local variable, etc.

There are already integrations that hackily add custom template functions ([example](https://github.com/PiotrMachowski/Home-Assistant-custom-components-Custom-Templates), [example](https://github.com/PiotrMachowski/Home-Assistant-custom-components-Saver#using-saved-values-in-templates)); can we please reopen this discussion?

**Reply by allenporter** [reply]
I think more discussion of the concrete use cases would help, given this has pretty large implications.

Can you describe more about the higher level or end user problems being solved here? e.g. what are the expensive operations, or recursive properties needed for, or what are the series of dates about. The technical reasons above are more details about why templates are proposed as a solution but not really exploring the bigger picture alternatives.

I'm a little unfamiliar with the full use cases, but it looks like getting an iterator or range of dates, but i can't tell.

Regarding computation: (1) How expensive is expensive? (2)  request options or different service calls can be used to return specific fields to avoid unnecessary computation, if this is really really expensive

**Comment by SLaks** [comment]
Can we revisit this?

**Reply by allenporter** [reply]
Given the comment above: Have you evaluated service return values as an alternative solution for the problem this would solve?

**Reply by SLaks** [reply]
Yes; see my reply above

**Reply by allenporter** [reply]
Ah I missed that since it wasn't threaded under the initial response
